File: dep/frontend/src/action/add_node_response.ts
import * as Graphflo from "graphflo";
import * as Guifast from "guifast_shared";

export namespace AddNodeResponse {
    const str = "add_node_response";
    export const id = Guifast.makeStrId(Graphflo.String.module, str);

    export interface Action extends Guifast.Action {
        readonly index: number;
        readonly node_desc: number;
        readonly input_connectors: Array<number>;
        readonly output_connectors: Array<number>;
    }

    export const make = (
        index: number,
        nodeDesc: number,
        inputConnectors: Array<number>,
        outputConnectors: Array<number>
    ): Action => {
        return {
            type: id,
            index: index,
            node_desc: nodeDesc,
            input_connectors: inputConnectors,
            output_connectors: outputConnectors
        };
    };
}

File: dep/backend/src/action/add_node_response.rs
use libflo_action_std::{ Action, action_str_id, NumberOrString };
use string;

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct AddNodeResponse<'a, 'b> {
    #[serde(rename = "type")]
    pub action_type: NumberOrString,
    pub index: usize,
    pub node_desc: usize,
    pub input_connectors: &'a Vec<usize>,
    pub output_connectors: &'b Vec<usize>
}

impl <'a, 'b> AddNodeResponse<'a, 'b> {
    pub fn new(index: usize, node_desc: usize, input_connectors: &'a Vec<usize>, output_connectors: &'b Vec<usize>) -> Self {
        AddNodeResponse {
            action_type: action_str_id(string::module(), string::add_node_response_action()),
            index: index,
            node_desc: node_desc,
            input_connectors: input_connectors,
            output_connectors: output_connectors
        }
    }
}

impl <'a, 'b> Action for AddNodeResponse<'a, 'b> {
    fn get_type(&self) -> &NumberOrString {
        &self.action_type
    }
}
